dev/src/document.ts

```diff
@@ -381,11 +381,26 @@ export class DocumentSnapshot<T = DocumentData> {
       return undefined;
     }
 
-    const obj: DocumentData = {};
-    for (const prop of Object.keys(fields)) {
-      obj[prop] = this._serializer.decodeValue(fields[prop]);
+    // We only want to use the converter and create a new QueryDocumentSnapshot
+    // if a converter has been provided.
+    if (this.ref._converter !== defaultConverter) {
+      const ref = new DocumentReference(this.ref.firestore, this.ref._path);
```

Comment:
I don't think you need to use the builder here since we are inside the DocumentSnapshot. Something like this would work:

```
      const untypedReference = new DocumentReference(this.ref.firestore, this.ref._path);
      return this.ref._converter.fromFirestore(
        new QueryDocumentSnapshot<DocumentData>(untypedReference, this._fieldsProto,
          this.readTime, this.createTime, this.updateTime)
      );
```

Reply:
done